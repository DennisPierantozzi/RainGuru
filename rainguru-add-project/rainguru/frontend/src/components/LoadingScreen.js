import React, {Component} from "react";

export default class LoadingScreen extends Component {

    // id for the set interval to rotate between the images
    static loadingIntervalId;

    // images for the loading screen
    loadingImages = ["../../static/images/loading1.webp", "../../static/images/loading2.webp", "../../static/images/loading3.webp"];

    /**
     * Sets the loading screen in the starting state.
     * @param props contains optional properties of the loading screen.
     */
    constructor(props) {
        super(props);

        // state variables
        this.state = {
            loadingImagesCount: 3,
            activeImage: this.loadingImages[0],
            activeImageIndex: 0
        }

        // start the interval to change between the 3 loading screens
        this.startLoading();
    }

    /**
     * Sets an interval to update the currently visible loading image to the next image.
     */
    startLoading() {
        // set the interval to every 300ms calculate the next index and set the image using the index
        LoadingScreen.loadingIntervalId = setInterval(function() {
            const newIndex = (this.state.activeImageIndex + 1) % this.state.loadingImagesCount;
            this.setState({activeImage: this.loadingImages[newIndex], activeImageIndex: newIndex});
        }.bind(this), 300);
    }

    /**
     * Renders the loading screen images.
     * @returns the rendered loading screen image.
     */
    render() {
        return <div id="loadingScreen" data-testid="loadingScreen"><img src={this.state.activeImage}
            alt={"Loading screen image " + this.state.activeImageIndex}/></div>;
    }
}