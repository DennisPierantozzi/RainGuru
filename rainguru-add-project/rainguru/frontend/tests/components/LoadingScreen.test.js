import React from "react";
import {render, screen} from "@testing-library/react";
import renderer from "react-test-renderer";
import "@testing-library/jest-dom/extend-expect";
import LoadingScreen from "../../src/components/LoadingScreen";

// disable certain warnings
const warnings = console.error.bind(console.error)
beforeAll(() => { console.error = errormessage => {
    !errormessage.toString().includes("Warning: Can't call \%s on a component that is not yet mounted") && warnings(errormessage)
}})
afterAll(() => { console.error = warnings })

describe("Loading screen tests", () => {
    it("matches snapshot", () => {
        const state = renderer.create(<LoadingScreen/>).toJSON();
        expect(state).toMatchSnapshot();
    });

    it("renders loading screen", () => {
        const ls = new LoadingScreen();
        render(ls.render());
        const element = screen.getByTestId("loadingScreen");

        expect(element).toBeTruthy();
        expect(LoadingScreen.loadingIntervalId).toBeDefined();
        expect(ls.loadingImages).toStrictEqual(["../../static/images/loading1.png",
            "../../static/images/loading2.png", "../../static/images/loading3.png"]);
        expect(ls.state.loadingImagesCount).toBe(3);
        expect(ls.state.activeImage).toBe("../../static/images/loading1.png");
        expect(ls.state.activeImageIndex).toBe(0);
    });

    it("goes to next loading frame correctly", () => {
        const ls = new LoadingScreen();
        // clear the current interval for refreshing the images, so it can be started again and the method can be tested
        clearInterval(LoadingScreen.loadingIntervalId);
        const spy = jest.spyOn(ls, "setState");

        // use fake timers to test the switching between loading images
        jest.useFakeTimers();

        // call startLoading to start the loading again
        ls.startLoading();

        // advance the timer to run the content of the set interval
        jest.advanceTimersByTime(300);
        expect(spy).toBeCalledTimes(1);
        expect(spy).toHaveBeenCalledWith({activeImage: "../../static/images/loading2.png", activeImageIndex: 1});

        // return the timers to normal just in case
        jest.useRealTimers();
    });

    it("goes to next loading frame correctly when it has to loop", () => {
        const ls = new LoadingScreen();
        // clear the current interval for refreshing the images, so it can be started again and the method can be tested
        clearInterval(LoadingScreen.loadingIntervalId);
        const spy = jest.spyOn(ls, "setState");

        // use fake timers to test the switching between loading images
        jest.useFakeTimers();

        // setup to loop the images
        ls.state.activeImage = "../../static/images/loading3.png";
        ls.state.activeImageIndex = 2;
        expect(ls.state.activeImage).toBe("../../static/images/loading3.png");
        expect(ls.state.activeImageIndex).toBe(2);

        // call startLoading to start the loading again
        ls.startLoading();

        // advance the timer to run the content of the set interval
        jest.advanceTimersByTime(300);
        expect(spy).toBeCalledTimes(1);
        expect(spy).toHaveBeenCalledWith({activeImage: "../../static/images/loading1.png", activeImageIndex: 0});

        // return the timers to normal just in case
        jest.useRealTimers();
    });
});
