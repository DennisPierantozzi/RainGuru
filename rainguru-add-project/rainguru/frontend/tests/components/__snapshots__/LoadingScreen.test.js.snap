// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading screen tests matches snapshot 1`] = `
<div
  data-testid="loadingScreen"
  id="loadingScreen"
>
  <img
    alt="Loading screen image 0"
    src="../../static/images/loading1.png"
  />
</div>
`;
